src/ui/public/visualize/visualization.js

```diff
@@ -17,137 +17,68 @@
-
-        $scope.vis.initialized = true;
-
-        const render$ = Rx.Observable.create(observer => {
-          $scope.$on('render', () => {
-            observer.next({
-              vis: $scope.vis,
-              visData: $scope.visData,
-              container: getVisContainer(),
-            });
-          });
-        }).pipe(
-          share()
-        );
-
-        const success$ = render$.pipe(
-          tap(() => {
-            dispatchRenderStart($el[0]);
-          }),
-          filter(({ vis, visData, container }) => vis && vis.initialized && container && (!vis.type.requiresSearch || visData)),
-          debounceTime(100),
-          switchMap(async ({ vis, visData, container }) => {
-            vis.size = [container.width(), container.height()];
-            const status = getUpdateStatus(vis.type.requiresUpdateStatus, $scope);
-            const renderPromise = visualization.render(visData, status);
-            $scope.$apply();
-            return renderPromise;
-          })
-        );
-
-        const requestError$ = render$.pipe(
-          filter(({ vis }) => vis.requestError)
-        );
-
-        const renderSubscription = Rx.merge(success$, requestError$)
-          .subscribe(() => {
-            dispatchRenderComplete($el[0]);
-          });
-
-        $scope.$on('$destroy', () => {
-          visualization.destroy();
-          renderSubscription.unsubscribe();
-        });
-
-        // Listen on changes to trigger a render if listenOnChange is not false
-        // i.e. true or has not been used (undefined in that case)
-        if ($scope.listenOnChange !== false) {
-          const onChangeListener = () => {
-            $scope.$emit('render');
-          };
-
-          $scope.$watchGroup(['visData', 'vis.params'], onChangeListener);
-
-          const resizeChecker = new ResizeChecker($el);
-          resizeChecker.on('resize', () => {
-            $scope.$emit('render');
-          });
-
-          $scope.uiState.on('change', onChangeListener);
-          $scope.$on('$destroy', () => {
-            resizeChecker.destroy();
-            $scope.uiState.off('change', onChangeListener);
-          });
-        }
-
-        function jQueryGetter(selector) {
-          return function () {
-            const $sel = $el.find(selector);
-            if ($sel.length) return $sel;
-          };
-        }
-      }
+import React, { Component } from 'react';
+import { VisualizationNoResults, VisualizationChart } from './components';
+
+const _showNoResultsMessage = (vis, visData) => {
+  const requiresSearch = _.get(vis, 'type.requiresSearch');
+  const isZeroHits = _.get(visData, 'hits.total') === 0;
+  const shouldShowMessage = !_.get(vis, 'params.handleNoResults');
+
+  return Boolean(requiresSearch && isZeroHits && shouldShowMessage);
+};
+
+export class Visualization extends Component {
+  constructor(props) {
+    super(props);
+    props.vis.initialized = true;
```

Comment:
I don't think it's a good idea to mutate props. We don't want any side effects on the props object passed.
I have to inspect better what is inside the vis object, but seems that if someone update the props of Visualization Component it will rerender loosing the mutated property on the constructor.

Reply:
this is a legacy code and will be removed as soon as we reactify <visualize> directive as well